render.c

```diff
@@ -88,11 +93,22 @@ static int render_notification(cairo_t *cairo, struct mako_state *state,
 		offset_x = style->margin.left;
 	}
 
+	double text_x = style->padding.left;
+#ifdef SHOW_ICONS
+	struct mako_icon icon;
+	if (style->show_icons) {
+		icon = get_icon(icon_path, style->max_icon_size);
```

Comment:
This should probably be done in `handle_notify`, rather than pulling the icon from disk every frame. The loaded image would then get cached in the `mako_notification`. This will end up reading the image even if icons are disabled, but I think this is a fair trade to avoid needing to scatter logic around other places to check if it gets changed by criteria and load the image then.

Reply:
This is also good preparation for adding support for icons via `icon_data` later on.